### ISSUE TITLE
FrechetDistance crash with large linestrings

### ISSUE THREAD

**Issue by Lgum** [pygeos/pygeos#439]
When calculating the Frechet distance between long linestrings (both ca. 8km in length), the function crashes and everything with it.

`from pygeos import Geometry, frechet_distance

ls1 = Geometry('<really long linestring>')
ls2 = Geometry('<another really long linestring>')
dist = frechet_distance(ls1, ls2)
print(dist)`

To give a frame of reference: This also happens with Spatialite so I assume it is an underlying issue with Geos, which might have a memory issue. Postgis, with a Geos 3.8.0 installation can handle this though. 

The linestrings can be found in this file:
[crash.txt](https://github.com/pygeos/pygeos/files/7587166/crash.txt)

Tested with Python 3.9.5 64Bit on Windows 10, pygeos 0.11.1 installed from pypi

**Comment by brendan-ward** [comment]
I'm not clear on what you mean by crashes.  Does the process exit with a segfault?  I'm assuming you omitted the WKT geometry from your example, since `Geometry('')` fails with an error.

I did not encounter an error testing with the latest version of GEOS (latest GEOS and pygoes built locally, on MacOS), using the test file you provided:

```
>>> import pandas as pd
>>> df = pd.read_csv("~/Downloads/crash.txt", delimiter=";")
>>> import pygeos as pg
>>> l = pg.from_wkt(df.ls.values)
>>> l
array([<pygeos.Geometry LINESTRING Z (7.62 47.507 314.8, 7.62 47.507 314.8, 7.62 47...>,
       <pygeos.Geometry LINESTRING Z (7.62 47.507 317.8, 7.62 47.507 317.8, 7.62 47...>],
      dtype=object)
>>> pg.length(l)
array([0.0957573 , 0.09529885])
>>> pg.get_num_coordinates(l)
array([2785, 2770], dtype=int32)
>>> pg.frechet_distance(l[0], l[1])
0.0002499039815609772
```

This should not be based on memory limits, these geometries are not that big.

I don't remember what version of GEOS is bundled into pygeos 0.11.1 on PyPI.  Could you run:
```
>>> import pygeos
>>> pygeos.geos_version
(3, 11, 0)
```

**Comment by jorisvandenbossche** [comment]
> I'm not clear on what you mean by crashes. Does the process exit with a segfault? I'm assuming you omitted the WKT geometry from your example, since Geometry('') fails with an error.

There is a txt file attached with WKT representations.

However, my the latest pygeos from conda-forge, I also can't reproduce this (using GEOS 3.10.0). Also in another environment with GEOS 3.9.1 I get a distance as result (in both cases I am using Linux (Ubuntu)).

**Comment by Lgum** [comment]
In my case, the python interpreter crashes with, presumably, a seg-fault, as soon as that function is executed:

`Process finished with exit code -1073741571 (0xC00000FD)`

The version output I get from pygeos.geos_version is:

`(3, 10, 0)`

Maybe it is hitting a recursion limit?

**Comment by caspervdw** [comment]
Both conda forge and or wheelbuilder use MSVC version 14.

Compare: https://github.com/pygeos/pygeos/runs/4264227956?check_suite_focus=true

With:
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=408813&view=logs&jobId=a70f640f-cc53-5cd3-6cdc-236a1aa90802&j=a70f640f-cc53-5cd3-6cdc-236a1aa90802&t=f5d15007-a01c-5ad8-c9ce-4d519d3b275f

**Comment by caspervdw** [comment]
I am leaning towards closing this issue as it seems to be an issue with GEOS, not with pygeos. Especially because it also happens with spatialite. @Lgum are you ok with this and can you raise the issue instead at the GEOS github repository?

**Comment by Lgum** [comment]
It is most likely a geos issue. I suspect that GEOS is using a recursive algorithm for the Frechet distance and in doing so, hits the stack-limit on windows, which, per default is 1 MB, in contrast to a default of 8 MB on unix systems. Sadly, I haven't been able to test this yet - building pygeos without libs hits a funky cython error on Windows. 

I could try to specifically set the stack size with a c-linker-flag. I'll see if I can make it work and will raise an issue at the geos repository. So it's fine by me if you close the issue.

**Comment by caspervdw** [comment]
Building on windows tends to give funky errors, yes. We do have a working script that builds on windows and some guidelines in the docs that may be of help:

https://github.com/pygeos/pygeos/blob/master/ci/install_geos.cmd

https://pygeos.readthedocs.io/en/stable/installation.html#installation-with-custom-geos-libary

Note that the last step “be sure that the GEOS dlls are available on your PATH” should be “copy the GEOS dlls into the package dir or you python dll dir” if you are using Python >=3.8.

If you keep hitting the error, please share it!

**Comment by Lgum** [comment]
I managed to build pygeos without binaries. However, what I had to do was:

- Create new venv
- Install numpy (and possibly cython) with wheels
- Then install pygeos with:
   - `python -m pip install pygeos --no-binary :all: --no-build-isolation`
   - While having set the include and lib dirs beforehand
- Afterwards I was able to copy the dll's into the python directory.

Sadly: It still crashed :)

I then created a small c++ program and tried to calculate the frechet distance there. The result:

- Crash when building GEOS without any special flags and building the program without special flags
- No crash when building GEOS without any special flags but specifying /F 8388608 and possibly also putting the stack reserve size to 8388608 and stack commit size to 16384 in the program, that links to the geos dll.

So the conclusion seems to be, that the program, that links to the dll needs to be build with the increased stack-size flags - not so much the dll?

I may put this also to the geos github, but at the moment I'm a bit clueless, how to proceed :)

**Comment by caspervdw** [comment]
Thanks for this thorough investigation! If you have the time, we could integrate your findings in the documentation on how to install pygeos from source on windows.

Could you explain to me what you did to make this work? What do you mean precisely by “ the program, that links to the dll”? Do you mean that GEOS (geos*.dll) should be built with increased stack size?

**Comment by Lgum** [comment]
Just tell me what you need for the documentation. It could take a while (all this is just a hobby of mine ;) ) but I can surely give you a step-by-step instruction how I built pygeos from source.

As for what I did to avoid the stackoverflow: I built GEOS normally (basically just downloaded the source, opened it in cmake-gui, generated the visual studio makefiles and built it without any special flags). Afterwards I set the linker flag in my own c++ project accordingly (stack reserve size = 8 mib). I assume that the program, that calls geos, is responsible for the stack - not the geos lib itself.

I suppose that if I build spatialite the same way, it will also stop crashing. As for how to build pygeos to avoid this, I honestly don't know. I am not familiar with the python build process and how it interacts with the stack sizes. Maybe with the threading module?

It's possible that it will be fixed in 3.10.2 - I posted an issue and it got labeled as a bug. So maybe you don't have to worry about this :) => https://github.com/libgeos/geos/issues/516

**Comment by caspervdw** [comment]
Now the issue was raised upstream, this can be closed. Concerning the documentation; there actually is already something in. 

> Get a C compiler applicable to your Python version (https://wiki.python.org/moin/WindowsCompilers)
Download and install a GEOS binary (https://trac.osgeo.org/osgeo4w/)
Set GEOS_INCLUDE_PATH and GEOS_LIBRARY_PATH environment variables (see below for notes on GEOS discovery)
Run pip install pygeos --no-binary
Make sure the GEOS .dll files are available on the PATH
